### PR TITLE
WSPdkB8v: Make form multi-part

### PR DESCRIPTION
**Pull request by kerrr**
This commit makes the onboarding form multi-part, allowing the user to
skip parts that are not required for their environment. Specifically:
* Allows user to specify an existing service config to copy
* If reusing an existing MSA, can just provide it's entity id
* Does not require MSA to be configured
* Skips integration test config if request of production access

### CODE REVIEW

**Comment by kerrr** [comment]
[Change the connecting form](https://trello.com/c/WSPdkB8v/233-change-the-connecting-form)

File: app/models/onboarding_form.rb [hunk 1]
```diff
@@ -62,50 +57,101 @@ class OnboardingForm
...
+  end
+
+  def validate_certificate(field_name)
+    begin
+      certificate = convert_value_to_x509_certificate(self.send(field_name))
+    rescue
+      errors[field_name] << "is malformed"
+    end
+    validate_certificate_issuer(certificate, field_name)
+  end
```
Comment (brenetic) [inline app/models/onboarding_form.rb]: Could use ensure here (don't need the begin/end):

```
def validate_certificate(field_name)
  certificate = convert_value_to_x509_certificate(self.send(field_name))
rescue
  errors[field_name] << "is malformed"
ensure
  validate_certificate_issuer(certificate, field_name)
end
```

Though if the `OpenSSL::X509::Certificate.new(Base64.decode64(cert_value))` errors might be good to do `validate_certificate_issuer(certificate, field_name) unless certificate.nil?`

File: app/views/environment_config_form/environment_config_form.html.erb [hunk 2]
```diff
@@ -18,6 +18,12 @@
   </div>
 <% end %>
 
+<div style="border: solid 1px red"><ul>
```
Comment (jakubmiarka) [inline app/views/environment_config_form/environment_config_form.html.erb]: debugging leftover?

File: app/controllers/environment_config_form_controller.rb [hunk 3]
```diff
@@ -38,6 +39,11 @@ def submit
 
   private
 
+  def render_part(part)
```
Comment (jakubmiarka) [inline app/controllers/environment_config_form_controller.rb]: is this used anywhere?

File: app/controllers/environment_config_form_controller.rb [hunk 4]
```diff
@@ -1,12 +1,14 @@
...
   def start
-    @onboarding_form = OnboardingForm.new({'environment_access' => 'integration-access-request'})
+    @onboarding_form = OnboardingForm.new({})
+    @onboarding_form.options=OnboardingOptions.new(session[:options])
     render 'environment_config_form'
   end
 
   def submit
     @onboarding_form = OnboardingForm.new(get_onboarding_form_params)
+    @onboarding_form.options=OnboardingOptions.new(session[:options])
```
Comment (jakubmiarka) [inline app/controllers/environment_config_form_controller.rb]:
```suggestion
    @onboarding_form.options = OnboardingOptions.new(session[:options])
```

File: app/controllers/environment_config_form_controller.rb [hunk 5]
```diff
@@ -1,12 +1,14 @@
 class EnvironmentConfigFormController < ApplicationController
 
   def start
-    @onboarding_form = OnboardingForm.new({'environment_access' => 'integration-access-request'})
+    @onboarding_form = OnboardingForm.new({})
+    @onboarding_form.options=OnboardingOptions.new(session[:options])
```
Comment (jakubmiarka) [inline app/controllers/environment_config_form_controller.rb]:
```suggestion
    @onboarding_form.options = OnboardingOptions.new(session[:options])
```

File: Gemfile.lock [hunk 6]
```diff
@@ -67,8 +68,8 @@ GEM
     faraday (0.13.1)
       multipart-post (>= 1.2, < 3)
     ffi (1.9.18)
-    globalid (0.3.7)
-      activesupport (>= 4.1.0)
+    globalid (0.4.2)
+      activesupport (>= 4.2.0)
     govuk_elements_rails (3.0.2)
```
Comment (jakubmiarka) [inline Gemfile.lock]: could we update this as well, while we're at it? there are some bug fixes (especially for Firefox)

File: app/views/environment_config_form/environment_config_form_integration.html.erb [hunk 7]
```diff
@@ -0,0 +1,122 @@
+<% content_for(
+       :page_title,
+       "#{'Error - ' if (@error || @onboarding_form.errors.any?)}Verify Environment Configuration - GOV.UK"
+   )
+%>
+
+<% if @error %>
```
Comment (jakubmiarka) [inline app/views/environment_config_form/environment_config_form_integration.html.erb]: are we doing any validation checks here? we might not need this (same with other views)

File: app/views/environment_config_form/environment_config_form_contact.html.erb [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    <%= f.hidden_field :user_account_middle_name%>
+    <%= f.hidden_field :user_account_surname%>
+    <%= f.hidden_field :user_account_dob%>
+    <%= f.hidden_field :user_account_current_address%>
+    <%= f.hidden_field :user_account_address_history%>
+    <%= f.hidden_field :user_account_cycle_3%>
+    <%= f.hidden_field :cycle3_attribute_name%>
+
+    <%= f.hidden_field :stub_idp_username%>
+    <%= f.hidden_field :stub_idp_password%>
```
Comment (jakubmiarka) [inline app/views/environment_config_form/environment_config_form_contact.html.erb]: all the hidden fields can go now, no?

Comment (jakubmiarka) [reply app/views/environment_config_form/environment_config_form_contact.html.erb]: in fact, the whole file?

File: app/views/environment_config_options/environment_config_options.html.erb [hunk 9]
```diff
@@ -0,0 +1,27 @@
+
+<div class="inner-block">
+  <%= form_for @onboarding_options, :url => { :action => "option_#{@onboarding_options.part}" }, :html => { :id => 'onboarding', :novalidate => 'novalidate', :multipart => true } do |f| %>
+
+    <%= session[:options].to_s %>
```
Comment (jakubmiarka) [inline app/views/environment_config_options/environment_config_options.html.erb]: debugging leftover

File: Gemfile.lock (hunk 6 above)
Comment (kerrr) [reply Gemfile.lock]: I updated to latest (3.1.3). It depends on `govuk_frontend_toolkit (>= 6.0.2)` which I also updated to. That broke the build, so I rolled back to investigate at another time.

File: app/models/onboarding_form.rb (hunk 1 above)
Comment (kerrr) [reply app/models/onboarding_form.rb]: I reworked this here in a different way.

File: app/models/onboarding_form_service.rb [hunk 10]
```diff
@@ -132,7 +138,7 @@ def delete_config_files(files)
 
     private
 
-    def value_or_default(value, default = '-')
+    def value_or_default(value, default = '-- Not Provided --')
```
Comment (kerrr) [inline app/models/onboarding_form_service.rb]: Having just the dash was interpreted in zendesk as markdown for the preceding line being a heading. This solves the problem.

File: Gemfile.lock (hunk 6 above)
Comment (jakubmiarka) [reply Gemfile.lock]: ok, no worries. let's do it in another PR ;)

**Pull request merged**